addons/resource_distribution_controller.go

```diff
@@ -0,0 +1,224 @@
+		WithEventFilter(predicate.NewPredicateFuncs(
+			func(object client.Object) bool {
+				return object.GetNamespace() == r.CurrentNamespace
+			}),
+		).
+		Complete(r)
+}
+
+func removeFinalizerFromObject(ctx context.Context, cl client.Client, o client.Object, finalizer string) error {
+	if controllerutil.RemoveFinalizer(o, finalizer) {
+		if err := cl.Update(ctx, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addFinalizerToObject(ctx context.Context, cl client.Client, o client.Object, finalizer string) error {
+	if controllerutil.AddFinalizer(o, finalizer) {
+		if err := cl.Update(ctx, o); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ResourceDistributionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger.With("Request", req)
+	logger.Info("Distributing resources to all StorageConsumers.")
+
+	var err error
+
+	var addonDeletionlock corev1.ConfigMap
+	err = r.SpokeClient.Get(ctx, types.NamespacedName{Namespace: r.CurrentNamespace, Name: AddonDeletionlockName}, &addonDeletionlock)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = addFinalizerToObject(ctx, r.SpokeClient, &addonDeletionlock, ResourceDistributionFinalizer)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	addVRCToConsumers := []string{}
+	storageConsumerList := ocsv1alpha1.StorageConsumerList{}
+	templateList := templatev1.TemplateList{}
+
+	err = r.SpokeClient.List(ctx, &storageConsumerList, client.InNamespace(r.CurrentNamespace))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(storageConsumerList.Items) == 0 {
+		logger.Info("No StorageConsumer to add/remove resources.")
+		err = removeFinalizerFromObject(ctx, r.SpokeClient, &addonDeletionlock, ResourceDistributionFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	err = r.SpokeClient.List(ctx, &templateList, client.InNamespace(r.CurrentNamespace), client.MatchingLabels{
+		utils.CreatedByLabelKey: utils.CreatorMulticlusterOrchestrator,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(templateList.Items) == 0 {
+		logger.Info("No resources to distribute to StorageConsumers.")
+		err = removeFinalizerFromObject(ctx, r.SpokeClient, &addonDeletionlock, ResourceDistributionFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	var storageClientMapping *corev1.ConfigMap
+	storageClientMapping, err = utils.GetStorageClientMapping(ctx, r.SpokeClient, r.CurrentNamespace)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+
+	for k := range storageClientMapping.Data {
+		for consumerIndex := range storageConsumerList.Items {
+			if storageConsumerList.Items[consumerIndex].Status.Client.ID == k {
```

Comment:
client resource under StorageConsumer Status is now a pointer, this code might panic if we don't check for != nil